File: labs/py/02.py
from math import sqrt

def calcDistance(points_pair):
    p1, p2 = points_pair
    d1 = p1[0] - p2[0]
    d2 = p1[1] - p2[1]
    return sqrt(d1 ** 2 + d2 ** 2)

def pointsToEdges(points):
    n = len(points)
    if n < 2: return []
    edges = [(points[0], points[n - 1])]
    for i in range(0, n - 1):
        edges.append((points[i], points[i + 1]))
    return edges

def calcPerimeter(edges):
    return sum(map(calcDistance, edges))

def getPoints():
    points = []
    while True:
        line = input("Enter point coordinates: ")
        if not line: break
        x, y = tuple(map(float, line.split()))
        points.append((x, y))
    return points

points    = getPoints()
edges     = pointsToEdges(points)
perimeter = calcPerimeter(edges)
print("Perimeter:", perimeter)

File: labs/hs/02.hs
import System.IO ( stdout, hFlush )

type Point = (Double, Double)
type Edge  = (Point, Point)

calcDistance :: Edge -> Double 
calcDistance ((x1, y1), (x2, y2))
    = sqrt $ (x1 - x2) ** 2 + (y1 - y2) ** 2

pointsToEdges :: [Point] -> [Edge]
pointsToEdges ps
    | length ps < 2 = []
    | otherwise     = (head ps, last ps) : helper ps
    where
    helper [p1, p2]   = [(p1, p2)]
    helper (p1:p2:ps) = (p1, p2) : helper (p2:ps)

calcPerimeter :: [Edge] -> Double
calcPerimeter = sum . map calcDistance

getPoints :: IO [Point]
getPoints = do
    line <- do
        putStr "Enter point coordinates: "
        hFlush stdout >> getLine
    if not $ null line then do
        let (x:y:_) = map read $ words line
        ((x, y):) <$> getPoints
    else return []

main :: IO ()
main = do
    p <- calcPerimeter . pointsToEdges <$> getPoints
    putStrLn $ "Perimeter: " ++ show p